File: pure_functions/pure-function.js
const PI = 3.14;

/* Not pure: it uses the global constant */
const calculateArea = (radius) => radius * radius * PI;

/* Pure: it uses only parameters passed to the function */
const calculateArea = (radius, PI) => radius * radius * PI;


var counter = 1;

var increaseCounter = function() {
  counter++;
  return counter;
}

increaseCounter(); // 2
counter; // 2


var counter = 1;

var increaseCounter = function(counter) {
  return counter + 1;
}

increaseCounter(counter); // 2
counter; // 1

File: pure_functions/pure-function.clj
;; Not a pure function because it uses an external global object
(def PI 3.14)

(defn calculate-area
  [radius]
  (* radius radius PI))

(calculate-area 10) ;; returns 314.0

;; pure function because it uses only parameters passed as arguments
(def PI 3.14)

(defn calculate-area
  [radius, PI]
  (* radius radius PI))

(calculate-area 10 PI) ;; returns 314.0

;; Not pure: it uses the global variable and modify the variable
(def counter 1)

(defn increase-counter
  [value]
  (def counter (inc value))
  counter)

(increase-counter counter)

;; pure function: returns the value increased by 1
(def counter 1)

(defn increase-counter
  [value]
  (inc value))

(increase-counter counter) ;; 2
counter ;; 1

;; impure function: reads an external file
(defn characters-counter
  [text]
  (str "Character count: " (count text)))

(defn analyze-file
  [filename]
  (characters-counter (slurp filename)))

(analyze-file "test.txt")

;; impure function: has random number generator
(defn year-end-evaluation
  []
  (if (> (rand) 0.5)
    "You get a raise!"
    "Better luck next year!"))

(defn increment-numbers
  [numbers]
  (map inc numbers))

(= [2 3 4 5 6] (increment-numbers [1 2 3 4 5]))
